File: app/promptfoo-sidecar/src/protocol.ts
export type EvalMode = "performance" | "trigger";

export type EvalAssertion = {
  type: "equals" | "contains" | "javascript";
  value: string;
};

export type EvalCase = {
  id: string;
  prompt: string;
  expected?: string;
  shouldTrigger?: boolean;
  assertions: EvalAssertion[];
};

export type EvalCandidate = {
  id: string;
  label: string;
  description?: string;
};

export type EvalDescriptionCandidate = EvalCandidate & {
  rationale?: string;
  rank?: number | null;
};

export type EvalExecution = {
  caseId: string;
  candidateId: string;
  output: unknown;
};

export type EvalHistoryConfig = {
  configDir: string;
  persist?: boolean;
};

export type EvalHistoryMetadata = {
  source: "eval_workbench";
  pluginSlug: string;
  skillName: string;
  scenarioName: string;
  mode: EvalMode;
};

export type RunEvalRequest = {
  id: string;
  type: "run_eval";
  mode: EvalMode;
  skillName: string;
  pluginSlug: string;
  scenarioName: string;
  promptfooConfigDir: string;
  candidates: EvalCandidate[];
  cases: EvalCase[];
  executions: EvalExecution[];
};

export type ListHistoryRequest = {
  id: string;
  type: "list_history";
  promptfooConfigDir: string;
  pluginSlug: string;
  skillName: string;
  scenarioName?: string;
  mode: EvalMode;
  limit: number;
};

export type ReadHistoryRequest = {
  id: string;
  type: "read_history";
  promptfooConfigDir: string;
  runId: string;
};

export type SidecarRequest =
  | RunEvalRequest
  | ListHistoryRequest
  | ReadHistoryRequest;

export type SidecarEvent =
  | {
      id: string;
      type: "progress";
      completed: number;
      total: number;
      caseId?: string;
      candidateId?: string;
    }
  | { id: string; type: "result"; result: EvalRunResult }
  | { id: string; type: "result"; runs: PersistedEvalRun[] }
  | { id: string; type: "result"; run: PersistedEvalRun | null }
  | { id: string; type: "error"; message: string };

export type EvalCaseResult = {
  caseId: string;
  candidateId: string;
  passed: boolean;
  score: number;
  output: unknown;
  reason?: string;
};

export type PersistedScenarioSnapshotCase = {
  id: string;
  prompt: string;
  expected?: string;
  shouldTrigger?: boolean;
  assertions: EvalAssertion[];
  sortOrder: number;
};

export type PersistedScenarioSnapshot = {
  pluginSlug: string;
  skillName: string;
  scenarioName: string;
  mode: EvalMode;
  cases: PersistedScenarioSnapshotCase[];
};

export type EvalRunResult = {
  mode: EvalMode;
  total: number;
  passed: number;
  failed: number;
  results: EvalCaseResult[];
  history?: {
    persisted: boolean;
    configDir?: string;
    evalId?: string;
    metadata?: EvalHistoryMetadata;
  };
};

export type EvalHistoryListItem = {
  evalId: string;
  createdAt: number;
  description?: string;
  total: number;
  passed: number;
  failed: number;
  metadata: EvalHistoryMetadata;
};

export type EvalHistoryListResult = {
  items: EvalHistoryListItem[];
  limit: number;
  offset: number;
};

export type EvalHistoryCaseDetail = {
  caseId?: string;
  candidateId?: string;
  prompt?: string;
  testIdx: number;
  promptIdx: number;
  success: boolean;
  score: number;
  response?: unknown;
  error?: string;
  latencyMs?: number;
  cost?: number;
  failureReason?: string | number;
  gradingResult?: unknown;
  metadata?: Record<string, unknown>;
  providerId?: string;
  providerLabel?: string;
};

export type EvalHistoryEntry = EvalHistoryListItem & {
  config?: Record<string, unknown>;
  cases: EvalHistoryCaseDetail[];
};

export type EvalHistoryReadResult = {
  entry: EvalHistoryEntry;
};

export type PersistedEvalRun = {
  id: string;
  promptfooEvalId: string;
  pluginSlug: string;
  skillName: string;
  scenarioName: string;
  mode: EvalMode;
  status: "completed";
  summary: {
    total: number;
    passed: number;
    failed: number;
    passRate: number;
  };
  scenarioSnapshot?: PersistedScenarioSnapshot;
  createdAt: string;
  completedAt: string | null;
  results: EvalCaseResult[];
};

const EVAL_MODES = new Set<EvalMode>(["performance", "trigger"]);
const ASSERTION_TYPES = new Set<EvalAssertion["type"]>([
  "equals",
  "contains",
  "javascript",
]);

export function parseSidecarRequest(line: string): SidecarRequest {
  let parsed: unknown;
  try {
    parsed = JSON.parse(line);
  } catch (error) {
    throw new Error(`Invalid JSONL request: ${formatError(error)}`);
  }

  if (!isRecord(parsed)) {
    throw new Error("Sidecar request must be a JSON object");
  }

  if (parsed.type === "run_eval") {
    return validateRunEvalRequest(parsed);
  }
  if (parsed.type === "list_history") {
    return validateListHistoryRequest(parsed);
  }
  if (parsed.type === "read_history") {
    return validateReadHistoryRequest(parsed);
  }

  const requestType = typeof parsed.type === "string" ? parsed.type : "unknown";
  throw new Error(`Unsupported sidecar request type: ${requestType}`);
}

export function serializeSidecarEvent(event: SidecarEvent): string {
  return `${JSON.stringify(event)}\n`;
}

export function serializeSidecarRequest(request: SidecarRequest): string {
  return `${JSON.stringify(request)}\n`;
}

function validateRunEvalRequest(
  value: Record<string, unknown>,
): RunEvalRequest {
  const id = requireString(value.id, "id");
  const mode = requireEvalMode(value.mode);
  const skillName = requireString(value.skillName, "skillName");
  const pluginSlug = requireString(value.pluginSlug, "pluginSlug");
  const scenarioName = requireString(value.scenarioName, "scenarioName");
  const promptfooConfigDir = requireString(
    value.promptfooConfigDir,
    "promptfooConfigDir",
  );
  const candidates = requireArray(value.candidates, "candidates").map(
    validateCandidate,
  );
  const cases = requireArray(value.cases, "cases").map(validateCase);
  const executions = requireArray(value.executions, "executions").map(
    validateExecution,
  );
  const history =
    value.history === undefined
      ? undefined
      : validateHistoryConfig(value.history, "history");
  const descriptionCandidates =
    value.descriptionCandidates === undefined
      ? undefined
      : requireArray(
          value.descriptionCandidates,
          "descriptionCandidates",
        ).map((candidate, index) =>
          validateDescriptionCandidate(candidate, index),
        );

  return {
    id,
    type: "run_eval",
    mode,
    skillName,
    pluginSlug,
    scenarioName,
    promptfooConfigDir,
    candidates,
    cases,
    executions,
  };
}

function validateListHistoryRequest(
  value: Record<string, unknown>,
): ListHistoryRequest {
  const request: ListHistoryRequest = {
    id: requireString(value.id, "id"),
    type: "list_history",
    promptfooConfigDir: requireString(
      value.promptfooConfigDir,
      "promptfooConfigDir",
    ),
    pluginSlug: requireString(value.pluginSlug, "pluginSlug"),
    skillName: requireString(value.skillName, "skillName"),
    mode: requireEvalMode(value.mode),
    limit: requirePositiveInteger(value.limit, "limit"),
  };
  if (value.scenarioName !== undefined) {
    request.scenarioName = requireString(value.scenarioName, "scenarioName");
  }
  return request;
}

function validateReadHistoryRequest(
  value: Record<string, unknown>,
): ReadHistoryRequest {
  return {
    id: requireString(value.id, "id"),
    type: "read_history",
    promptfooConfigDir: requireString(
      value.promptfooConfigDir,
      "promptfooConfigDir",
    ),
    runId: requireString(value.runId, "runId"),
  };
}

function validateCandidate(value: unknown, index: number): EvalCandidate {
  if (!isRecord(value)) {
    throw new Error(`candidates[${index}] must be an object`);
  }

  const candidate: EvalCandidate = {
    id: requireString(value.id, `candidates[${index}].id`),
    label: requireString(value.label, `candidates[${index}].label`),
  };

  if (value.description !== undefined) {
    candidate.description = requireString(
      value.description,
      `candidates[${index}].description`,
    );
  }

  return candidate;
}

function validateDescriptionCandidate(
  value: unknown,
  index: number,
): EvalDescriptionCandidate {
  const candidate = validateCandidate(value, index);
  if (!isRecord(value)) {
    throw new Error(`descriptionCandidates[${index}] must be an object`);
  }

  const descriptionCandidate: EvalDescriptionCandidate = { ...candidate };
  if (value.rationale !== undefined) {
    descriptionCandidate.rationale = requireString(
      value.rationale,
      `descriptionCandidates[${index}].rationale`,
    );
  }
  if (value.rank !== undefined) {
    descriptionCandidate.rank = requireNullableInteger(
      value.rank,
      `descriptionCandidates[${index}].rank`,
    );
  }
  return descriptionCandidate;
}

function validateCase(value: unknown, index: number): EvalCase {
  if (!isRecord(value)) {
    throw new Error(`cases[${index}] must be an object`);
  }

  const testCase: EvalCase = {
    id: requireString(value.id, `cases[${index}].id`),
    prompt: requireString(value.prompt, `cases[${index}].prompt`),
    assertions: requireArray(
      value.assertions,
      `cases[${index}].assertions`,
    ).map((assertion, assertionIndex) =>
      validateAssertion(assertion, index, assertionIndex),
    ),
  };

  if (value.expected !== undefined) {
    testCase.expected = requireString(
      value.expected,
      `cases[${index}].expected`,
    );
  }
  if (value.shouldTrigger !== undefined) {
    testCase.shouldTrigger = requireBoolean(
      value.shouldTrigger,
      `cases[${index}].shouldTrigger`,
    );
  }

  return testCase;
}

function validateExecution(value: unknown, index: number): EvalExecution {
  if (!isRecord(value)) {
    throw new Error(`executions[${index}] must be an object`);
  }

  return {
    caseId: requireString(value.caseId, `executions[${index}].caseId`),
    candidateId: requireString(
      value.candidateId,
      `executions[${index}].candidateId`,
    ),
    output: value.output,
  };
}

function validateAssertion(
  value: unknown,
  caseIndex: number,
  assertionIndex: number,
): EvalAssertion {
  const field = `cases[${caseIndex}].assertions[${assertionIndex}]`;
  if (!isRecord(value)) {
    throw new Error(`${field} must be an object`);
  }

  if (
    typeof value.type !== "string" ||
    !ASSERTION_TYPES.has(value.type as EvalAssertion["type"])
  ) {
    throw new Error(`${field}.type must be equals, contains, or javascript`);
  }

  return {
    type: value.type as EvalAssertion["type"],
    value: requireString(value.value, `${field}.value`),
  };
}

function validateHistoryConfig(
  value: unknown,
  field: string,
): EvalHistoryConfig {
  if (!isRecord(value)) {
    throw new Error(`${field} must be an object`);
  }

  const history: EvalHistoryConfig = {
    configDir: requireString(value.configDir, `${field}.configDir`),
  };

  if (value.persist !== undefined) {
    history.persist = requireBoolean(value.persist, `${field}.persist`);
  }

  return history;
}

function validateHistoryFilter(value: unknown): ListEvalHistoryRequest["filter"] {
  if (!isRecord(value)) {
    throw new Error("filter must be an object");
  }

  const filter: ListEvalHistoryRequest["filter"] = {
    configDir: requireString(value.configDir, "filter.configDir"),
    pluginSlug: requireString(value.pluginSlug, "filter.pluginSlug"),
    skillName: requireString(value.skillName, "filter.skillName"),
  };

  if (value.scenarioName !== undefined) {
    filter.scenarioName = requireString(
      value.scenarioName,
      "filter.scenarioName",
    );
  }
  if (value.mode !== undefined) {
    filter.mode = requireEvalMode(value.mode);
  }
  if (value.limit !== undefined) {
    filter.limit = requirePositiveInteger(value.limit, "filter.limit");
  }
  if (value.offset !== undefined) {
    filter.offset = requireNonNegativeInteger(value.offset, "filter.offset");
  }

  return filter;
}

function requireEvalMode(value: unknown): EvalMode {
  if (typeof value === "string" && EVAL_MODES.has(value as EvalMode)) {
    return value as EvalMode;
  }

  throw new Error("mode must be performance or trigger");
}

function requireString(value: unknown, field: string): string {
  if (typeof value === "string" && value.length > 0) {
    return value;
  }

  throw new Error(`${field} must be a non-empty string`);
}

function requireBoolean(value: unknown, field: string): boolean {
  if (typeof value === "boolean") {
    return value;
  }

  throw new Error(`${field} must be a boolean`);
}

function requireArray(value: unknown, field: string): unknown[] {
  if (Array.isArray(value)) {
    return value;
  }

  throw new Error(`${field} must be an array`);
}

function requirePositiveInteger(value: unknown, field: string): number {
  if (typeof value === "number" && Number.isInteger(value) && value > 0) {
    return value;
  }

  throw new Error(`${field} must be a positive integer`);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function formatError(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: app/src-tauri/src/agents/promptfoo_sidecar/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum EvalMode {
    Performance,
    Trigger,
}

impl EvalMode {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Performance => "performance",
            Self::Trigger => "trigger",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum EvalAssertionType {
    Equals,
    Contains,
    Javascript,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EvalAssertion {
    #[serde(rename = "type")]
    pub assertion_type: EvalAssertionType,
    pub value: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EvalCase {
    pub id: String,
    pub prompt: String,
    #[serde(default)]
    pub expected: Option<String>,
    #[serde(default)]
    pub should_trigger: Option<bool>,
    #[serde(default)]
    pub assertions: Vec<EvalAssertion>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct EvalCandidate {
    pub id: String,
    pub label: String,
    #[serde(default)]
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EvalDescriptionCandidate {
    pub id: String,
    pub label: String,
    pub description: String,
    #[serde(default)]
    pub rationale: Option<String>,
    #[serde(default)]
    pub rank: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EvalExecution {
    pub case_id: String,
    pub candidate_id: String,
    pub output: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum RequestType {
    RunEval,
    ListHistory,
    ReadHistory,
}

fn default_run_eval_request_type() -> RequestType {
    RequestType::RunEval
}

fn default_list_history_request_type() -> RequestType {
    RequestType::ListHistory
}

fn default_read_history_request_type() -> RequestType {
    RequestType::ReadHistory
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RunEvalRequest {
    pub id: String,
    #[serde(rename = "type", default = "default_run_eval_request_type")]
    pub request_type: RequestType,
    pub mode: EvalMode,
    pub skill_name: String,
    pub plugin_slug: String,
    pub scenario_name: String,
    pub promptfoo_config_dir: String,
    pub candidates: Vec<EvalCandidate>,
    pub cases: Vec<EvalCase>,
    pub executions: Vec<EvalExecution>,
}

impl RunEvalRequest {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        id: impl Into<String>,
        mode: EvalMode,
        skill_name: impl Into<String>,
        plugin_slug: impl Into<String>,
        scenario_name: impl Into<String>,
        promptfoo_config_dir: impl Into<String>,
        candidates: Vec<EvalCandidate>,
        cases: Vec<EvalCase>,
        executions: Vec<EvalExecution>,
    ) -> Self {
        Self {
            id: id.into(),
            request_type: "run_eval",
            mode,
            skill_name: skill_name.into(),
            plugin_slug: plugin_slug.into(),
            scenario_name: scenario_name.into(),
            promptfoo_config_dir: promptfoo_config_dir.into(),
            candidates,
            cases,
            executions,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ListHistoryRequest {
    pub id: String,
    #[serde(rename = "type", default = "default_list_history_request_type")]
    pub request_type: RequestType,
    pub promptfoo_config_dir: String,
    pub plugin_slug: String,
    pub skill_name: String,
    #[serde(default)]
    pub scenario_name: Option<String>,
    pub mode: EvalMode,
    pub limit: i64,
}

impl ListHistoryRequest {
    pub fn new(
        id: impl Into<String>,
        promptfoo_config_dir: impl Into<String>,
        plugin_slug: impl Into<String>,
        skill_name: impl Into<String>,
        scenario_name: Option<String>,
        mode: EvalMode,
        limit: i64,
    ) -> Self {
        Self {
            id: id.into(),
            request_type: RequestType::ListHistory,
            promptfoo_config_dir: promptfoo_config_dir.into(),
            plugin_slug: plugin_slug.into(),
            skill_name: skill_name.into(),
            scenario_name,
            mode,
            limit,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ReadHistoryRequest {
    pub id: String,
    #[serde(rename = "type", default = "default_read_history_request_type")]
    pub request_type: RequestType,
    pub promptfoo_config_dir: String,
    pub run_id: String,
}

impl ReadHistoryRequest {
    pub fn new(
        id: impl Into<String>,
        promptfoo_config_dir: impl Into<String>,
        run_id: impl Into<String>,
    ) -> Self {
        Self {
            id: id.into(),
            request_type: RequestType::ReadHistory,
            promptfoo_config_dir: promptfoo_config_dir.into(),
            run_id: run_id.into(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ListEvalHistoryFilter {
    pub config_dir: String,
    pub plugin_slug: String,
    pub skill_name: String,
    #[serde(default)]
    pub scenario_name: Option<String>,
    #[serde(default)]
    pub mode: Option<EvalMode>,
    #[serde(default)]
    pub limit: Option<u32>,
    #[serde(default)]
    pub offset: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ListEvalHistoryRequest {
    pub id: String,
    #[serde(rename = "type")]
    pub request_type: &'static str,
    pub filter: ListEvalHistoryFilter,
}

impl ListEvalHistoryRequest {
    pub fn new(id: impl Into<String>, filter: ListEvalHistoryFilter) -> Self {
        Self {
            id: id.into(),
            request_type: "list_eval_history",
            filter,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ReadEvalHistoryRequest {
    pub id: String,
    #[serde(rename = "type")]
    pub request_type: &'static str,
    pub config_dir: String,
    pub eval_id: String,
}

impl ReadEvalHistoryRequest {
    pub fn new(
        id: impl Into<String>,
        config_dir: impl Into<String>,
        eval_id: impl Into<String>,
    ) -> Self {
        Self {
            id: id.into(),
            request_type: "read_eval_history",
            config_dir: config_dir.into(),
            eval_id: eval_id.into(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EvalCaseResult {
    pub case_id: String,
    pub candidate_id: String,
    pub passed: bool,
    pub score: f64,
    pub output: Value,
    #[serde(default)]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EvalRunHistory {
    pub persisted: bool,
    #[serde(default)]
    pub config_dir: Option<String>,
    #[serde(default)]
    pub eval_id: Option<String>,
    #[serde(default)]
    pub metadata: Option<EvalHistoryMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EvalRunResult {
    pub mode: EvalMode,
    pub total: u32,
    pub passed: u32,
    pub failed: u32,
    pub results: Vec<EvalCaseResult>,
    #[serde(default)]
    pub history: Option<EvalRunHistory>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EvalHistoryListItem {
    pub eval_id: String,
    pub created_at: i64,
    #[serde(default)]
    pub description: Option<String>,
    pub total: u32,
    pub passed: u32,
    pub failed: u32,
    pub metadata: EvalHistoryMetadata,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EvalHistoryListResult {
    pub items: Vec<EvalHistoryListItem>,
    pub limit: u32,
    pub offset: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EvalHistoryCaseDetail {
    #[serde(default)]
    pub case_id: Option<String>,
    #[serde(default)]
    pub candidate_id: Option<String>,
    #[serde(default)]
    pub prompt: Option<String>,
    pub test_idx: u32,
    pub prompt_idx: u32,
    pub success: bool,
    pub score: f64,
    #[serde(default)]
    pub response: Option<Value>,
    #[serde(default)]
    pub error: Option<String>,
    #[serde(default)]
    pub latency_ms: Option<f64>,
    #[serde(default)]
    pub cost: Option<f64>,
    #[serde(default)]
    pub failure_reason: Option<Value>,
    #[serde(default)]
    pub grading_result: Option<Value>,
    #[serde(default)]
    pub metadata: Option<Value>,
    #[serde(default)]
    pub provider_id: Option<String>,
    #[serde(default)]
    pub provider_label: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EvalHistoryEntry {
    #[serde(flatten)]
    pub header: EvalHistoryListItem,
    #[serde(default)]
    pub config: Option<Value>,
    pub cases: Vec<EvalHistoryCaseDetail>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EvalHistoryReadResult {
    pub entry: EvalHistoryEntry,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PersistedEvalRunSummary {
    pub total: u32,
    pub passed: u32,
    pub failed: u32,
    pub pass_rate: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PersistedEvalRun {
    pub id: String,
    pub promptfoo_eval_id: String,
    pub plugin_slug: String,
    pub skill_name: String,
    pub scenario_name: String,
    pub mode: EvalMode,
    pub status: String,
    pub summary: PersistedEvalRunSummary,
    #[serde(default)]
    pub scenario_snapshot: Option<serde_json::Value>,
    pub created_at: String,
    pub completed_at: Option<String>,
    pub results: Vec<EvalCaseResult>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(untagged)]
pub enum SidecarResultPayload {
    Eval { result: EvalRunResult },
    Runs { runs: Vec<PersistedEvalRun> },
    Run { run: Box<Option<PersistedEvalRun>> },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SidecarEvent {
    Progress {
        id: String,
        completed: u32,
        total: u32,
        #[serde(rename = "caseId", default)]
        case_id: Option<String>,
        #[serde(rename = "candidateId", default)]
        candidate_id: Option<String>,
    },
    Result {
        id: String,
        #[serde(flatten)]
        payload: SidecarResultPayload,
    },
    HistoryListResult {
        id: String,
        result: EvalHistoryListResult,
    },
    HistoryReadResult {
        id: String,
        result: EvalHistoryReadResult,
    },
    Error {
        id: String,
        message: String,
    },
}

pub fn serialize_request<T>(request: &T) -> Result<String, String>
where
    T: Serialize,
{
    serde_json::to_string(request)
        .map(|json| format!("{json}\n"))
        .map_err(|error| error.to_string())
}

pub fn parse_sidecar_event(line: &str) -> Result<SidecarEvent, String> {
    serde_json::from_str(line).map_err(|error| error.to_string())
}
